R/pkg/R/SQLContext.R

```diff
@@ -90,19 +96,31 @@ createDataFrame <- function(sqlContext, data, schema = NULL, samplingRatio = 1.0
       if (is.null(schema)) {
         schema <- names(data)
       }
-      n <- nrow(data)
-      m <- ncol(data)
-      # get rid of factor type
-      dropFactor <- function(x) {
+
+      # types that should be wrapped before
+      # mapply to avoid flattening
+      requires_wrapping <- c("list")
+
+      # get rid of factor type and adjust for lists
+      cleanCols <- function(x) {
         if (is.factor(x)) {
           as.character(x)
+        } else if(class(x) %in% requires_wrapping) {
```

Comment:
@zero323, I don't understand the meaning here. could you give me an example?

An experiment like:
```
mapply(list, c(1:2), list(list(3,4),list(5,6)), SIMPLIFY=FALSE)
```
works as expected. The combination of atomic vector and list works fine for mapply.

I would recommend to eliminate this code piece here so that this PR focus on performance improvement only.